src/helper/merge-html-plugin.ts

```diff
@@ -0,0 +1,162 @@
+  This plugin allows this, so that we can implement highlights with tooltips.
+
+  Taken from: https://github.com/highlightjs/highlight.js/issues/2889
+*/
+
+import { HighlightResult, HLJSPlugin } from 'highlight.js';
+
+export const mergeHTMLPlugin = (function () {
+  let originalStream: Event[];
+
+  function escapeHTML (value: string): string {
+    return value
+      .replace(/&/g, '&amp;')
+      .replace(/</g, '&lt;')
+      .replace(/>/g, '&gt;')
+      .replace(/"/g, '&quot;')
+      .replace(/'/g, '&#x27;');
+  }
+
+  /* plugin itself */
+
+  const mergeHTMLPlugin: HLJSPlugin = {
+    // preserve the original HTML token stream
+    'before:highlightElement': ({ el }: {el: Node}) => {
+      originalStream = nodeStream(el);
+    },
+    // merge it afterwards with the highlighted token stream
+    'after:highlightElement': ({ el, result, text }: {el: Element; result: HighlightResult; text: string}) => {
+      if (originalStream.length === 0) return;
+
+      const resultNode = document.createElement('div');
+      resultNode.innerHTML = result.value;
+      result.value = mergeStreams(originalStream, nodeStream(resultNode), text);
+      el.innerHTML = result.value;
+    }
+  };
+
+  /* Stream merging support functions */
+
+  interface Event {
+    event: 'start' | 'stop';
+    offset: number;
+    node: Node;
+  }
+
+  function tag (node: Node): string {
+    return node.nodeName.toLowerCase();
+  }
+
+  function nodeStream (node: Node): Event[] {
+    const result: Event[] = [];
+    (function _nodeStream (node, offset) {
+      for (let child = node.firstChild; child != null; child = child.nextSibling) {
+        if (child.nodeType === 3) {
+          offset += child.nodeValue?.length ?? 0;
+        } else if (child.nodeType === 1) {
+          result.push({
+            event: 'start',
+            offset,
+            node: child
+          });
+          offset = _nodeStream(child, offset);
+          // Prevent void elements from having an end tag that would actually
+          // double them in the output. There are more void elements in HTML
+          // but we list only those realistically expected in code display.
+          if (tag(child).match(/br|hr|img|input/) == null) {
+            result.push({
+              event: 'stop',
+              offset,
+              node: child
+            });
+          }
+        }
+      }
+      return offset;
+    })(node, 0);
+    return result;
+  }
+
+  function mergeStreams (original: Event[], highlighted: Event[], value: string): string {
+    let processed = 0;
+    let result = '';
+    const nodeStack = [];
+
+    function selectStream (): Event[] {
+      if ((original.length === 0) || (highlighted.length === 0)) {
+        return (original.length > 0) ? original : highlighted;
+      }
+      if (original[0].offset !== highlighted[0].offset) {
+        return (original[0].offset < highlighted[0].offset) ? original : highlighted;
+      }
+
+      /*
```

Comment:
We don't need code comments I believe.

Reply:
This plugin is taken from https://github.com/highlightjs/highlight.js/issues/2889. I think we should just keep the source as is, I just made it TypeScript compatible. I credited the source at the top.

Reply:
If we're cloning it into our repo instead of using it directly from the node_modules as a dependency, we don't need these comments. Please avoid it, instead just put a reference in the THIRD PARTY file.